### ISSUE TITLE
npm@5 errors under docker with NFS volume

### ISSUE THREAD

**Issue by mtibben** [npm/npm#17578]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

When running `npm install` in a docker container with a mounted volume located on a NFS filesystem, npm errors out with `npm ERR! Error: EPERM: operation not permitted, chown ...`. 

```
$ docker run -it --volume $(pwd):/usr/local/src --workdir /usr/local/src node:8 \
  npm install --save mkdirp
npm info it worked if it ends with ok
npm info using npm@5.0.3
npm info using node@v8.1.2
npm http fetch GET 200 https://registry.npmjs.org/mkdirp 24ms (from cache)
npm http fetch GET 200 https://registry.npmjs.org/minimist 17ms (from cache)
npm ERR! Cannot read property 'write' of null
npm WARN enoent ENOENT: no such file or directory, open '/usr/local/src/package.json'
npm WARN src No description
npm WARN src No repository field.
npm WARN src No README data
npm WARN src No license field.

npm ERR! path /usr/local/src/node_modules/.staging/minimist-e209f42f
npm ERR! code EPERM
npm ERR! errno -1
npm ERR! syscall chown
npm ERR! Error: EPERM: operation not permitted, chown '/usr/local/src/node_modules/.staging/minimist-e209f42f'
npm ERR!  { Error: EPERM: operation not permitted, chown '/usr/local/src/node_modules/.staging/minimist-e209f42f'
npm ERR!   cause:
npm ERR!    { Error: EPERM: operation not permitted, chown '/usr/local/src/node_modules/.staging/minimist-e209f42f'
npm ERR!      errno: -1,
npm ERR!      code: 'EPERM',
npm ERR!      syscall: 'chown',
npm ERR!      path: '/usr/local/src/node_modules/.staging/minimist-e209f42f' },
npm ERR!   errno: -1,
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'chown',
npm ERR!   path: '/usr/local/src/node_modules/.staging/minimist-e209f42f' }
npm ERR!
npm ERR! Please try running this command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2017-07-03T02_04_20_906Z-debug.log
```
For the full log see [2017-07-03T02_04_20_906Z-debug.log](https://gist.github.com/mtibben/58f1529a224853a0d8f26abefefc7530#file-2017-07-03t02_04_20_906z-debug-log)

Interestingly, when I create a non-root user, the command works correctly. Docker by default uses the root user inside a container, so it seems npm has some logic executing when running under the root user.

This problem did not exist in previous versions of npm

#### How can the CLI team reproduce the problem?

1. Enter a docker container with a mounted volume located on a NFS filesystem
2. Run `npm install --save mkdirp` from the volume directory

I've created a script to reproduce the problem on macOS https://gist.github.com/mtibben/58f1529a224853a0d8f26abefefc7530#file-npm-nfs-test-case-sh

### supporting information:

 - `npm -v` prints: `5.0.3`
 - `node -v` prints: `v8.1.2`
 - `npm config get registry` prints:
```
npm info it worked if it ends with ok
npm info using npm@5.0.3
npm info using node@v8.1.2
https://registry.npmjs.org/
npm info ok
```
 - Windows, OS X/macOS, or Linux?: macOS
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [x] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by Maxwell2022** [comment]
Not sure if it's related, but we cannot use NPM 5 with circleci: https://discuss.circleci.com/t/npm-install-error-that-isnt-reproducing-in-a-local-docker-container

**Comment by vektah** [comment]
Related to https://github.com/npm/npm/issues/3565, not sure why its closed though.

**Comment by mtibben** [comment]
Looks like this is fixed in npm 5.4
